python/lsst/cp/pipe/cpFlatNormTask.py

```diff
@@ -74,93 +165,207 @@ class CpFlatNormalizationTask(pipeBase.PipelineTask,
+
+        if self.config.level == 'DETECTOR':
+            for detId, det in enumerate(detSet):
+                for amp in camera[detId]:
+                    for expId, exp in enumerate(expSet):
+                        outputScales['expScale'][det][amp.getName()][exp] = expScales[expId]
+                outputScales['detScale'][det] = detScales[detId]
+        elif self.config.level == 'AMP':
+            for detId, det in enumerate(detSet):
+                for ampIdx, amp in enumerate(camera[detId]):
+                    for expId, exp in enumerate(expSet):
+                        outputScales['expScale'][det][amp.getName()][exp] = expScales[expId]
+                    detAmpId = detId + ampIdx
+                    outputScales['detScale'][det][amp.getName()] = detScales[detAmpId]
 
         return pipeBase.Struct(
             outputScales=outputScales,
         )
+
+    def measureScales(self, bgMatrix, bgCounts, iterations=10):
+        """Convert backgrounds to exposure and detector components.
+
+        Parameters
+        ----------
+        bgMatrix : `np.ndarray`
+            Input backgrounds indexed by exposure (axis=0) and
+            detector (axis=1).
+        bgCounts : `np.ndarray`
+            Input pixel counts used to in measuring bgMatrix, indexed
+            identically.
+        iterations : `int`
+            Number of iterations to use in decomposition.
+
+        Returns
+        -------
+        vectorE : `np.array`
+            Output E vector of exposure level scalings.
+        vectorG : `np.array`
+            Output G vector of detector level scalings.
+        bgModel : `np.ndarray`
+            Expected model bgMatrix values, calculated from E and G.
+
+        Notes
+        -----
+
+        The set of background measurements B[exposure, detector] of
+        flat frame data should be defined by a "Cartesian" product of
+        two vectors, E[exposure] and G[detector].  The E vector
+        represents the total flux incident on the focal plane.  In a
+        perfect camera, this is simply the sum along the columns of B
+        (np.sum(B, axis=0)).
+
+        However, this simple model ignores differences in detector
+        gains, the vignetting of the detectors, and the illumination
+        pattern of the source lamp.  The G vector describes these
+        detector dependent differences, which should be identical over
+        different exposures.  For a perfect lamp of unit total
+        intensity, this is simply the sum along the rows of B
+        (np.sum(B, axis=1)).
+
+        The algorithm here, from pipe_drivers/constructCalibs.py and
+        from there from Eugene Magnier/PanSTARRS [1]_, attempts to
+        iteratively solve this decomposition from initial "perfect" E
+        and G vectors.  The operation is performed in log space to
+        reduce the multiply and divides to linear additions and
+        subtractions.
```

Comment:
I think it's important to mention that the average of the G vector is constrained to unity by the algorithm (i.e., we're solving for the _relative_ gains, not absolute values).